NextJS Edge Runtime Support

### Describe the feature in detail (code, mocks, or screenshots encouraged)

Support for usage in NextJs Edge Runtime

### What type of pull request would this be?

Other

### Provide relevant links or additional information.

Usefull Docs: https://nextjs.org/docs/app/api-reference/edge

# Use Case
We're storing OAuth2 Client Credentials in Vault. Those credentials are used by next-auth api endpoints to allow users to log themself in and by the next-auth middleware to provide and update the user session. Right now we're using the fetch method provided by nextjs to load those secrets but we would love to use this package.

# Error
I run into this error when creating a new instance of `Client` on Edge Runtime:

```
⨯ Error: Cannot find module 'node:net': Unsupported external type Url for commonjs reference
    at <unknown> (file://C:\Users\<user>\IdeaProjects\frontend\.next\server\edge\chunks\ssr\bb6bd_undici_de3331._.js:8070:15)
    at Object.<anonymous> (file://C:\Users\<user>\IdeaProjects\frontend\.next\server\edge\chunks\ssr\bb6bd_undici_de3331._.js:8073:3)
    at [project]/node_modules/.pnpm/undici@6.19.8/node_modules/undici/lib/dispatcher/client.js [middleware] (ecmascript) (file://C:\Users\<user>\IdeaProjects\frontend\.next\server\edge\chunks\ssr\bb6bd_undici_de3331._.js:8525:3)
    at <unknown> (file://C:\Users\<user>\IdeaProjects\frontend\.next\server\edge\chunks\edge-wrapper_ae8a2c.js:535:27)
    at runModuleExecutionHooks (file://C:\Users\<user>\IdeaProjects\frontend\.next\server\edge\chunks\edge-wrapper_ae8a2c.js:577:9)
    at instantiateModule (file://C:\Users\<user>\IdeaProjects\frontend\.next\server\edge\chunks\edge-wrapper_ae8a2c.js:533:9)
    at getOrInstantiateModuleFromParent (file://C:\Users\<user>\IdeaProjects\frontend\.next\server\edge\chunks\edge-wrapper_ae8a2c.js:609:12)
    at commonJsRequire (file://C:\Users\<user>\IdeaProjects\frontend\.next\server\edge\chunks\edge-wrapper_ae8a2c.js:146:20)
    at Object.<anonymous> (file://C:\Users\<user>\IdeaProjects\frontend\.next\server\edge\chunks\ssr\bb6bd_undici_de3331._.js:21086:16)
    at [project]/node_modules/.pnpm/undici@6.19.8/node_modules/undici/index.js [middleware] (ecmascript) (file://C:\Users\<user>\IdeaProjects\frontend\.next\server\edge\chunks\ssr\bb6bd_undici_de3331._.js:21223:3)
```

--- a/.changeset/metal-games-beam.md
+++ b/.changeset/metal-games-beam.md
@@ -1,0 +1,7 @@
+---
+'@litehex/node-vault': major
+---
+
+Fix: Using the environment's global `fetch` API instead of `undici`.
+
+This change makes this library more compatible with JavaScript runtimes that do not support `Node.js` modules. If you want to use `undici` again,  use the [Custom Fetcher](https://github.com/shahradelahi/node-vault/wiki/Usage#custom-fetcher) feature.

--- a/package.json
+++ b/package.json
@@ -34,7 +34,6 @@
   "packageManager": "pnpm@8.15.7",
   "dependencies": {
     "p-safe": "^1.0.0",
-    "undici": "^6.19.2",
     "zod": "^3.23.8",
     "zod-request": "^0.2.2"
   },

--- a/package.json
+++ b/package.json
@@ -55,7 +55,8 @@
     "tsd": "^0.31.1",
     "tsup": "^8.1.0",
     "tsx": "^4.15.7",
-    "typescript": "^5.5.2"
+    "typescript": "^5.5.2",
+    "undici": "^6.19.8"
   },
   "type": "module",
   "engines": {

--- a/src/index.ts
+++ b/src/index.ts
@@ -1,10 +1,3 @@
-import { fetch } from 'undici';
-import { setGlobalFetch } from 'zod-request';
-
-setGlobalFetch(fetch);
-
-// -------------------------------
-
 export { Client } from '@/lib/client';
 export { generateCommand } from '@/utils/generate-command';
 

--- a/src/lib/client.ts
+++ b/src/lib/client.ts
@@ -1,4 +1,3 @@
-import type { RequestInit } from 'undici';
 import { z } from 'zod';
 
 import { Aws } from '@/engine/aws';

--- a/src/typings.ts
+++ b/src/typings.ts
@@ -1,7 +1,6 @@
 /* eslint-disable @typescript-eslint/ban-types */
 
 import { SafeReturn } from 'p-safe';
-import { RequestInit, Response } from 'undici';
 import type { z } from 'zod';
 import type { RequestSchema as ZodRequestSchema } from 'zod-request';
 

--- a/src/typings.ts
+++ b/src/typings.ts
@@ -8,7 +8,7 @@ import { VaultError } from '@/errors';
 import { ClientOptionsSchema } from '@/schema';
 
 export type ClientOptions = z.infer<typeof ClientOptionsSchema> & {
-  request?: Partial<RequestInit>;
+  request?: Partial<RequestInit & Record<string, unknown>>;
   fetcher?: Fetcher;
 };
 
@@ -19,7 +19,7 @@ export type RequestSchema = Omit<ZodRequestSchema, 'path' | 'body'> & {
 
 export type Fetcher = (input: any, init: any) => Promise<any>;
 
-export interface ExtendedRequestInit extends RequestInit {
+export interface ExtendedRequestInit extends RequestInit, Record<string, unknown> {
   strictSchema?: boolean;
 }
 

--- a/src/utils/generate-command.ts
+++ b/src/utils/generate-command.ts
@@ -1,7 +1,6 @@
 import omit from 'lodash/omit';
 import pick from 'lodash/pick';
 import { SafeReturn, trySafe } from 'p-safe';
-import { fetch, type RequestInit } from 'undici';
 import { z } from 'zod';
 import { generateRequest, ZodRequestInit, ZodResponse } from 'zod-request';
 

--- a/tests/integration.test.ts
+++ b/tests/integration.test.ts
@@ -1,6 +1,6 @@
 import { expect } from 'chai';
 import { expectType } from 'tsd';
-import { ProxyAgent } from 'undici';
+import { fetch, ProxyAgent } from 'undici';
 import { z } from 'zod';
 
 import { Client, generateCommand, VaultError } from '@/index';
@@ -129,6 +129,7 @@ describe('node-vault', () => {
       expect(url).to.be.instanceof(URL);
       expect(url.toString()).to.equal('http://127.0.0.1:8200/v1/secret-path/test');
       used = true;
+      // @ts-expect-error Init type has some missing properties
       return fetch(url, init);
     };
 
@@ -181,6 +182,7 @@ describe('node-vault', () => {
     }
 
     const agent = new ProxyAgent(HTTP_PROXY);
+    vc.fetcher = fetch;
 
     const status = await vc.sealStatus(undefined, { dispatcher: agent });
 

--- a/tests/utils.ts
+++ b/tests/utils.ts
@@ -1,7 +1,6 @@
 import { execSync } from 'node:child_process';
 import { accessSync } from 'node:fs';
 import { resolve } from 'node:path';
-import { expect } from 'chai';
 
 import { Client } from '@/index';
 
@@ -25,7 +24,7 @@ export async function createInstance(unsealed: boolean = true): Promise<{
     secret_shares: 1,
     secret_threshold: 1
   });
-  expect(error, error?.message).be.be.undefined;
+  if (error) throw error;
 
   const { keys, root_token } = data!;
 

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -1,6 +1,6 @@
 {
   "compilerOptions": {
-    "lib": ["ESNext"],
+    "lib": ["ESNext", "DOM", "DOM.Iterable"],
     "module": "esnext",
     "target": "esnext",
     "moduleResolution": "node",

